### PR TITLE
MV3 click to load support

### PR DESCRIPTION
**Pull request by ladamski**
https://app.asana.com/0/72649045549333/1202839073279061/f

### CODE REVIEW

File: lib/tds.js [hunk 1]
```diff
@@ -278,9 +297,9 @@ async function generateDNRRulesForTrackerEntry (
         if (surrogate) {
             resourceTypes = ['script']
             if (!supportedSurrogateScripts.has(surrogate)) {
-                // Block requests if an unsupported surrogate script rule
-                // matches.
-                ruleAction = 'block'
+                // we should ignore actions for missing surrogates, otherwise breakage can result?
```
Comment (kzar) [inline lib/tds.js:300]: Yea, I did it that way initially, but then I found the expected behavior was to block. [See this reference test](https://github.com/duckduckgo/privacy-reference-tests/blob/main/tracker-radar-tests/TR-domain-matching/domain_matching_tests.json#L260).

Comment (ladamski) [reply lib/tds.js]: 👍  fixed

File: test/tds.js [hunk 2]
```diff
@@ -1551,4 +1551,273 @@ describe('generateTdsRuleset', () => {
...
+            },
+            {
+                rule: 'default-custom\\.invalid\\/block-ctl-yt-2',
+                action: 'block-ctl-yt',
+                exceptions: { types: ['script'] }
+            }
+        ])
+
+        await rulesetEqual(blockList, isRegexSupportedTrue, null, {
+            expectedRuleset: [{
```
Comment (ladamski) [inline test/tds.js:1613]: Note that currently these tests capture the actual behavior, will spend a bit more time tomorrow thinking through whether the expected behavior

File: lib/tds.js [hunk 3]
```diff
@@ -418,6 +450,8 @@ function finalizeDNRRulesAndLookup (startingRuleId, dnrRules) {
  *   The generated Tracker Blocking declarativeNetRequest ruleset.
  * @property {object} matchDetailsByRuleId
  *   Rule ID -> match details.
+ * @property {object} customActionRules
```
Comment (kzar) [inline lib/tds.js:453]: Please could you rename this `inverseRulesByCustomRuleAction` and expand the comment (maybe "List of inverse rules to add for each custom rule action"?).

File: lib/tds.js [hunk 4]
```diff
@@ -290,19 +309,30 @@ async function generateDNRRulesForTrackerEntry (
...
+            matchCase,
+            requestDomains,
+            excludedInitiatorDomains,
+            resourceTypes
+        })
+
+        if (customRuleAction) {
+            // if a custom action, map the action to the specified default
+            // and add its inverse to the list of custom action rules
+            const inverseRule = { customAction: trackerEntryRules[i].action, ...newRule }
```
Comment (kzar) [inline lib/tds.js:328]: I would prefer to call `generateDNRRule` again here, but with the inverse action. Otherwise, we're sharing some Objects between the two rules which could lead to confusing bugs where one is mutated and the other rule changes in an unexpected way.

File: lib/tds.js [hunk 5]
```diff
@@ -505,14 +540,23 @@ async function generateTdsRuleset (
     const dnrRules = []
     for (const [trackerDomain, trackerEntry] of Object.entries(tds.trackers)) {
         const requestDomains = requestDomainsByTrackerDomain.get(trackerDomain)
+        if (typeof tds.entities[trackerEntry.owner.name] === 'undefined') {
```
Comment (kzar) [inline lib/tds.js:543]: Not sure that this check is necessary, or why we need to add it now? Could you explain?

**Review by kzar — CHANGES_REQUESTED** [review]
@ladamski I've taken a first pass, though I've not looked through the test changes yet!

File: lib/tds.js [hunk 6]
```diff
@@ -190,17 +196,19 @@ async function generateDNRRulesForTrackerEntry (
...
     const dnrRules = []
+    const customActions = []
 
     if (priority > MAXIMUM_SUBDOMAIN_PRIORITY) {
         throw new Error('Too many tracker entries for domain: ' + trackerDomain)
     }
 
     const defaultAction = normalizeAction(trackerEntry.default)
-
+    const customDefaultAction = supportedCustomActions[defaultAction]
```
[the remark this answers is not in the document]
Comment (ladamski) [reply lib/tds.js:206]: Wait did we?  I honestly don't remember that and didn't see that in the TD either.  FWIW we just discussed this in https://app.asana.com/0/1171671347221384/1203389689265368 and sounds like we'll handle this just in rules. So I'll remove this for now as its not strictly needed, but seems like it introduces a bit of an odd inconsistency?

File: lib/tds.js [hunk 7]
```diff
@@ -7,6 +7,12 @@ const {
     processRegexTrackerRule
 } = require('./utils')
 
+// enumerates the supported custom actions, and the DNR actions they should map to
```
[the remark this answers is not in the document]
Comment (ladamski) [reply lib/tds.js:10]: 👍

File: lib/tds.js [hunk 8]
```diff
@@ -7,6 +7,12 @@ const {
...
+const supportedCustomActions = {
+    'block-ctl-yt': { action: 'block', inverseAction: 'allow' },
```
[the remark this answers is not in the document]
Comment (ladamski) [reply lib/tds.js:12]: Well its a mapping, so it maps one action to another?  Not sure what else to call it that wouldn't be more confusing.

File: lib/tds.js [hunk 9]
```diff
@@ -527,8 +571,7 @@ async function generateTdsRuleset (
             dnrRules.push(rule)
         }
     }
-
-    return finalizeDNRRulesAndLookup(startingRuleId, dnrRules)
+    return { ...finalizeDNRRulesAndLookup(startingRuleId, dnrRules), customActionRules }
```
[the remark this answers is not in the document]
Comment (ladamski) [reply lib/tds.js:574]: I took a look though I'm not clear where the `supported-surrogates-input.json` is supposed to come from?    It doesn't work with `mapping.json` since it doesn't parse as a Set?

File: lib/tds.js [hunk 10]
```diff
@@ -236,14 +249,20 @@ async function generateDNRRulesForTrackerEntry (
         } = trackerEntryRules[i]
 
         ruleAction = normalizeAction(ruleAction, 'block')
+        const customRuleAction = supportedCustomActions[ruleAction]
 
         // So far, only 'block' and 'allow' (aka 'ignore') actions are supported
```
[the remark this answers is not in the document]
Comment (ladamski) [reply lib/tds.js:254]: 👍

File: lib/tds.js (hunk 5 above)
Comment (ladamski) [reply lib/tds.js:547]: Because just below it (tries to) deference `tds.entities[trackerEntry.owner.name].domains`, and from testing with locally generated TDS this would sometimes fail.   Perhaps this is cleaned up during the TDS release process, but regardless it should be handled.

File: lib/tds.js (hunk 3 above)
Comment (ladamski) [reply lib/tds.js]: Sure done, though to be pedantic I called it `inverseCustomActionRules`, since its an array not a map. :)

File: lib/tds.js [hunk 11]
```diff
@@ -324,7 +355,8 @@ async function generateDNRRulesForTrackerEntry (
         }
     }
 
-    return removeRedundantDNRRules(dnrRules)
+    const trackerRules = removeRedundantDNRRules(dnrRules)
+    return { trackerRules, customActions }
```
[the remark this answers is not in the document]
Comment (ladamski) [reply lib/tds.js]: 👍

File: lib/tds.js (hunk 4 above)
Comment (ladamski) [reply lib/tds.js]: Done

Comment (ladamski) [reply lib/tds.js]: Done!

File: test/tds.js (hunk 2 above)
Comment (ladamski) [reply test/tds.js:1606]: Ok I think the tests are worth a look now too

File: lib/tds.js [hunk 12]
```diff
@@ -290,19 +310,29 @@ async function generateDNRRulesForTrackerEntry (
...
+            resourceTypes
+        }
+        const newRule = generateDNRRule(newRuleTemplate)
+
+        if (customRuleAction) {
+            // if a custom action, map the action to the specified default
+            // and add its inverse to the list of custom action rules
+            const inverseRule = { ...newRuleTemplate }
+            inverseRule.actionType = customRuleAction.inverseAction
+            newRule[inverseCustomActionSymbol] = { customAction: trackerEntryRules[i].action, ...generateDNRRule(inverseRule) }
```
Comment (kzar) [inline lib/tds.js:332]: How about we assign the custom action here, then let the finalize step do the rest?
```suggestion
            newRule[customActionSymbol] = customRuleAction
```

File: lib/tds.js [hunk 13]
```diff
@@ -311,6 +341,7 @@ async function generateDNRRulesForTrackerEntry (
             // declarativeNetRequest rules with a 'block'/'redirect' action of
             // the same priority.
             // See https://developer.chrome.com/docs/extensions/reference/declarativeNetRequest/#matching-algorithm
+
```
Comment (kzar) [inline lib/tds.js:344]: Nit: Unrelated change?

File: lib/tds.js [hunk 14]
```diff
@@ -348,9 +381,13 @@ function finalizeDNRRulesAndLookup (startingRuleId, dnrRules) {
         // just add the rule to the ruleset and match details to the lookup now.
         if (!rule.condition.requestDomains ||
             rule.priority !== BASELINE_PRIORITY ||
-            rule.action === 'redirect') {
+            rule.action === 'redirect' || rule[inverseCustomActionSymbol]) {
```
Comment (kzar) [inline lib/tds.js:384]: Nit: Would you mind moving this onto a new line and also adding a note about this case in the above comment? Just so that it's clear the inverse rule is something used by Click to Load and not related to surrogate script redirection.

File: lib/tds.js [hunk 15]
```diff
@@ -47,6 +61,7 @@ const MAXIMUM_REGEX_RULES = 900
 // declarativeNetRequest rule to aid the creation of the trackerDomainByRuleId
 // lookup.
 const trackerDomainSymbol = Symbol('trackerDomain')
+const inverseCustomActionSymbol = Symbol('customAction')
```
Comment (kzar) [inline lib/tds.js:64]: Nit: `customAction` ?

Also, could you expand on the above comment to briefly mention about custom actions?

File: lib/tds.js [hunk 16]
```diff
@@ -324,14 +355,16 @@ async function generateDNRRulesForTrackerEntry (
...
+    const trackerRules = removeRedundantDNRRules(dnrRules)
+    return { trackerRules }
 }
 
 function finalizeDNRRulesAndLookup (startingRuleId, dnrRules) {
     const ruleIdByByStringifiedDNRRule = new Map()
     const requestDomainsByRuleId = new Map()
     const trackerDomainsByRuleId = new Map()
     const matchDetailsByRuleId = {}
+    const inverseCustomActionRules = []
```
Comment (kzar) [inline lib/tds.js:367]: Should this not be an Object that maps each customAction its array of inverse rules?
```suggestion
    const inverseRulesByCustomAction = {}
```

File: lib/tds.js (hunk 5 above)
Comment (kzar) [reply lib/tds.js:547]: Fair enough, but could you get rid of the warning and add a comment to mention that sometimes the block list is malformed, so we should take care to skip tracker entries that don't have a corresponding entity entry?

File: lib/tds.js [hunk 17]
```diff
@@ -324,14 +355,16 @@ async function generateDNRRulesForTrackerEntry (
         }
     }
 
-    return removeRedundantDNRRules(dnrRules)
+    const trackerRules = removeRedundantDNRRules(dnrRules)
```
Comment (kzar) [inline lib/tds.js:358]: I think this change is unnecessary now?

**Review by kzar — CHANGES_REQUESTED** [review]
Looks way better @ladamski :+1: 

I think my main suggestion is that we just store the custom action with each rule (instead of the inverse rule and the custom action), then let the finalize step take care of generating the inverse rules. That way, we have the custom action and the inverse rule to hand at the point we're creating the lookup, which avoids doing such a dance.

Otherwise just nits I think!

Also, sorry I ran out of time to look at the tests, but will do that when I take my next pass.

File: lib/tds.js (hunk 12 above)
Comment (ladamski) [reply lib/tds.js:332]: I actually disagree.  This method is `generateDNRRulesForTrackerEntry`, and the new rule is generated right above via `newRuleTemplate`, and so its inverse should be generated here as well.

Otherwise I'd have to copy or recreate that template in the finalize step, which means we're muddling where rule generation happens and run the risk of drift over time where the inverse may no longer have the same properties as the original.  Or I go back to cloning the generated DNR rule in finalize.  

Already rewrote this once, I think we should leave it be unless there's a specific issue to fix?

File: lib/tds.js (hunk 17 above)
Comment (ladamski) [reply lib/tds.js]: 👍

File: lib/tds.js (hunk 14 above)
Comment (ladamski) [reply lib/tds.js]: 👍

File: lib/tds.js (hunk 12 above)
Comment (kzar) [reply lib/tds.js:334]: Well, I think it would make more sense to let finalize step take care of it. To my mind, this is kind of meta-data for the rule, that it has this custom action that also needs to be handled. The rule itself and the custom action is all that's needed to generate the inverse rule (or rules for different kinds of custom actions in the future). It's not known if the inverse rule (or whatever results from other sorts of custom actions) is required until the finalize step.

I don' t think implementing it the way I'm suggesting would be too hard. Here we'd do something like:
```javascript
let customRuleAction = null
if (supportedCustomActions.has(ruleAction)) {
    customRuleAction = ruleAction
    ruleAction = supportedCustomActions.get(ruleAction).action
}

...

if (customRuleAction) {
    newRule[customActionSymbol] = customRuleAction
}
```

and then in the finalize function, we'd do something like:

```javascript
const inverseRulesByCustomAction = new Map()

...

const customAction = rule[customRuleActionSymbol]
delete  rule[customRuleActionSymbol]
if (customAction) {
    const customActionDetails = supportedCustomActions.get(customAction)
    
    if (customActionDetails.inverseAction) {
        const inverseRule = JSON.parse(JSON.stringify(rule))
        storeInLookup(inverseRulesByCustomAction, customAction, [inverseRule])
    }
}

...

return { ruleset, matchDetailsByRuleId, inverseRulesByCustomAction }

```

File: lib/tds.js (hunk 4 above)
Comment (kzar) [reply lib/tds.js]: Thanks, yep either would work. To be fair, that would probably make more sense too (see below)!

File: lib/tds.js (hunk 16 above)
Comment (kzar) [reply lib/tds.js:367]: (This is covered in my above comment, resolving.)

File: lib/tds.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (kzar) [reply lib/tds.js]: Fair enough, I'll resolve this since neither of us can think of a better name.

File: lib/tds.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (kzar) [reply lib/tds.js]: Sorry, I need to update the README about that. For testing, I have a `supported-surrogates.json` file that contains `["analytics.js"]`.

File: lib/tds.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (kzar) [reply lib/tds.js]: Well, my understanding was that we did. The logic was that tracker rules with an unknown action are skipped, but default actions are always standard (block/ignore). That's since if a tracker rule is skipped, matching can fall back to other rules for the tracker, or the default action. But falling back to an unknown default action isn't going to work.

File: lib/tds.js (hunk 12 above)
Comment (ladamski) [reply lib/tds.js:334]: I think we need to do a call about this, I'm available until about 9am your time or again your evening.

**Review by kzar — APPROVED** [review]
